### DISCUSSION TITLE
Gramine v1.5, seeing a suplpemental data size mismatch

### DISCUSSION THREAD

**Discussion by mgild** [gramineproject/gramine#1474] [General]
On verifying a quote in gramine:v1.5 with azure:

`Warning: Quote supplemental data size is different between DCAP QVL and QvE, please make sure you installed DCAP QVL and QvE from same release`

Installation steps in dockerfile:

```
FROM gramineproject/gramine:v1.5

RUN curl -fsSLo /usr/share/keyrings/microsoft.asc https://packages.microsoft.com/keys/microsoft.asc
RUN echo "deb [arch=amd64 signed-by=/usr/share/keyrings/microsoft.asc] https://packages.microsoft.com/ubuntu/20.04/prod focal main" | \
    tee /etc/apt/sources.list.d/msprod.list

RUN apt-get update
RUN apt-get upgrade -y
RUN apt install -y az-dcap-client
```
Im producing this quote in a container which is running graminev1.4. Do containers that produce the quotes need to match the gramine version verifying the quotes?

**Comment by kailun-qin** [comment]
Thanks for the report!

> Warning: Quote supplemental data size is different between DCAP QVL and QvE, please make sure you installed DCAP QVL and QvE from same release

But this is basically a warning from DCAP QVL which indicates a supplemental data version mismatch between QVL and QvE (see https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/e7604e02331b3377f3766ed3653250e03af72d45/QuoteVerification/dcap_quoteverify/sgx_dcap_quoteverify.cpp#L411-L421)?

I assume that you're running the quote verification on an SGX platform. Would you pls kindly check the installed versions of the related SGX libs in your env (via `sudo apt list --installed | grep sgx`)?

**Comment by mgild** [comment]
so I have 

```
libsgx-ae-epid/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-ae-id-enclave/unknown,now 1.16.100.2-focal1 amd64 [installed,upgradable to: 1.17.100.4-focal1]
libsgx-ae-le/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-ae-pce/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-ae-qe3/unknown,now 1.16.100.2-focal1 amd64 [installed,upgradable to: 1.17.100.4-focal1]
libsgx-ae-qve/unknown,now 1.16.100.2-focal1 amd64 [installed,upgradable to: 1.17.100.4-focal1]
libsgx-aesm-ecdsa-plugin/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-aesm-epid-plugin/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-aesm-launch-plugin/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-aesm-pce-plugin/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-aesm-quote-ex-plugin/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-dcap-default-qpl/unknown,now 1.17.100.4-focal1 amd64 [installed]
libsgx-dcap-quote-verify/unknown,now 1.16.100.2-focal1 amd64 [installed,upgradable to: 1.17.100.4-focal1]
libsgx-enclave-common/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-launch/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
libsgx-pce-logic/unknown,now 1.16.100.2-focal1 amd64 [installed,upgradable to: 1.17.100.4-focal1]
libsgx-qe3-logic/unknown,now 1.16.100.2-focal1 amd64 [installed,upgradable to: 1.17.100.4-focal1]
libsgx-urts/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
sgx-aesm-service/unknown,now 2.19.100.3-focal1 amd64 [installed,upgradable to: 2.20.100.4-focal1]
```


But I am doing quote verification via a copy of this module:

https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs


Should I use a certain version of this rust module (some git tag I should check out before copying to my code?)

@kailun-qin

**Reply by kailun-qin** [reply]
> But I am doing quote verification via a copy of this module:
> https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/master/QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs

Thanks for the info! Pls note that in `sgx_qve_header.h`, `MAX_SA_LIST_SIZE` has been updated from 160 bytes to 320 bytes in DCAP 1.17 release (see https://github.com/intel/SGXDataCenterAttestationPrimitives/commit/0443ae263818a78afeeaf50ca29fc2cff02cb829#diff-9f7c89f78a91c61460125dad11cf736a89fcc16e86f33a8517c0c435321f820f) which'll change the quote supplemental data size (see https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/e7604e02331b3377f3766ed3653250e03af72d45/QuoteVerification/QvE/Include/sgx_qve_header.h#L128). Since the header'll be used for bindgen during the Rust module build (specifically `intel-tee-quote-verification-sys` -- the dependency of `intel-tee-quote-verification-rs`), this can possibly lead to the incompatible results observed.

> Should I use a certain version of this rust module (some git tag I should check out before copying to my code?)

Yes, pls try checking out tag `DCAP_1.16` (https://github.com/intel/SGXDataCenterAttestationPrimitives/tree/DCAP_1.16) which matches your installed QVE version and give it another shot. Thanks!

**Reply by mgild** [reply]
hm @kailun-qin still seems problematic on 1.16

**Reply by mgild** [reply]
https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/dcap_1.16_reproducible/SampleCode/RustQuoteVerificationSample/src/main.rs#L172-L187

Code I am using

**Reply by mgild** [reply]
`tee_get_supplemental_data_version_and_size` provides a size of `336`
`mem::size_of::<sgx_ql_qv_supplemental_t>() as u32` == `496`

**Reply by mgild** [reply]
I do seem to need to install `libsgx-dcap-quote-verify-dev` myself given its not included in gramine 1.5.  Is there a version of this I should use?

**Reply by mgild** [reply]
Ah from https://download.01.org/intel-sgx/sgx_repo/ubuntu/pool/main/devel/libsgx-dcap-quote-verify-dev/ i believe im using 1.17 here, I will try 1.16

**Reply by mgild** [reply]
I believe same for [libsgx-dcap-default-qpl_1.16.100.2-focal1_amd64.deb](https://download.01.org/intel-sgx/sgx_repo/ubuntu/pool/main/libs/libsgx-dcap-default-qpl/libsgx-dcap-default-qpl_1.16.100.2-focal1_amd64.deb)

**Reply by mgild** [reply]
That fixed it :)

**Comment by mgild** [comment]
Thanks so much @kailun-qin!

Also a bit of an aside, I want to help people test gramine related code on any platform. Is there any way people can run gramine-direct on mac m2 chips? (Or anything arm compatible just to test log output of logs of a gramine direct run)

**Reply by kailun-qin** [reply]
> Is there any way people can run gramine-direct on mac m2 chips? (Or anything arm compatible just to test log output of logs of a gramine direct run)

Unfortunately, Gramine currently only works on the x86_64 architecture :).

**Reply by mkow** [reply]
And ppc64, although not officially :) (see https://github.com/stefanberger/gramine4ppc64)

**Comment by mgild** [comment]
Hm got it. For gramine-direct what’s the exact reason there, isn’t everything simulated? Running in a docker environment and using Rosetta is fine for me. I’m making a generic dev platform using gramine and would love to reduce the barrier for entry of mac users just for testing before launching their applications.

**Reply by mkow** [reply]
Rosetta may actually work I think, but I never tried it and I'm not a macOS user, so no idea ¯\\\_(ツ)\_\/¯

**Reply by dimakuv** [reply]
> Hm got it. For gramine-direct what’s the exact reason there, isn’t everything simulated?

Gramine has assumptions on some architecture-specific properties, like the general-purpose registers (GPRs), how the stack grows, how the exception-handling frames are generated, how the extended features (XSAVE area in x86-64) are saved/restored. You can find these snippets in the Gramine codebase here and there, under the `arch/` subdirectories.

It is definitely possible to write an ARM-based version of Gramine, but it will surely require some effort.

> Running in a docker environment and using Rosetta is fine for me.

That should actually work. Did you try it?

**Reply by mgild** [reply]
Yes @dimakuv , @mkow   I see
```
error: Gramine with Linux backend requires XSAVE CPU instruction(s). Please upgrade your hardware.
error: PAL failed verify_hw_requirements() failed
```

**Reply by dimakuv** [reply]
@mgild Very interesting. This is the source code where the error happens: https://github.com/gramineproject/gramine/blob/4de6aa5fb939a129d3498e8fee432d2873924b4c/pal/src/host/linux-sgx/host_main.c#L1147-L1155

According to https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits, the `xsave` bit 26 stands for:

> Extensible processor state save/restore: XSAVE, XRSTOR, XSETBV, XGETBV instructions

So, surprisingly, your Mac with Rosetta either doesn't emulate the aforementioned x86 instructions, or it doesn't expose that it supports these instructions in CPUID leaf 0x1.

**Reply by mgild** [reply]
If there is any workaround, I would love to get my users testing through the gramine stack @dimakuv

**Reply by dimakuv** [reply]
@mgild [This page](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment) says that Rosetta doesn't support some newer processor features, though it talks mostly about Intel AVX. I wonder if Rosetta also doesn't support Extended State Enumeration at all. Unfortunately, Gramine relies on the XSAVE feature, and I don't know how to work around this.

**Reply by mgild** [reply]
Is there a way for just gramine-direct to work?

**Reply by dimakuv** [reply]
@mgild Gramine-direct also uses mbedTLS (e.g. for Encrypted Files), and the XSAVE is required for mbedTLS. So no.
